astropy/table/table.py

```diff
@@ -3712,24 +3719,27 @@ def _rows_equal(self, other):
         if isinstance(other, Table):
             other = other.as_array()
 
-        if self.has_masked_columns:
-            if isinstance(other, np.ma.MaskedArray):
-                result = self.as_array() == other
-            else:
-                # If mask is True, then by definition the row doesn't match
-                # because the other array is not masked.
-                false_mask = np.zeros(1, dtype=[(n, bool) for n in self.dtype.names])
-                result = (self.as_array().data == other) & (self.mask == false_mask)
-        else:
-            if isinstance(other, np.ma.MaskedArray):
+        try:
+            self_is_masked = self.has_masked_columns
+            other_is_masked = isinstance(other, np.ma.MaskedArray)
+            if (self_is_masked + other_is_masked) == 1:
+                # remap variables to a and b where a is masked and b isn't
+                a, b = (
+                    (self.as_array(), other)
+                    if self_is_masked
+                    else (other, self.as_array())
+                )
+
                 # If mask is True, then by definition the row doesn't match
                 # because the other array is not masked.
-                false_mask = np.zeros(1, dtype=[(n, bool) for n in other.dtype.names])
-                result = (self.as_array() == other.data) & (other.mask == false_mask)
+                false_mask = np.zeros(1, dtype=[(n, bool) for n in a.dtype.names])
+                return (a.data == b) & (a.mask == false_mask)
```

Comment:
Can you revert to setting `result` and returning that at the end? Overall I prefer this pattern and try to avoid early return if it doesn't require contorted logic or a giant `if` block.

This is one thing in astropy that each subpackage has its own style and it helps to maintain that style FWIW.  :smile: